polaris-react/src/components/UnstyledButton/UnstyledButton.tsx

```diff
@@ -13,92 +13,107 @@ export interface UnstyledButtonProps extends BaseButton {
+    {
+      id,
+      children,
+      className,
+      url,
+      external,
+      download,
+      submit,
+      disabled,
+      loading,
+      pressed,
+      accessibilityLabel,
+      role,
+      ariaControls,
+      ariaExpanded,
+      ariaDescribedBy,
+      ariaChecked,
+      onClick,
+      onFocus,
+      onBlur,
+      onKeyDown,
+      onKeyPress,
+      onKeyUp,
+      onMouseEnter,
+      onTouchStart,
+      ...rest
+    }: UnstyledButtonProps,
+    ref,
+  ) => {
+    let buttonMarkup;
+    const localRef: React.RefObject<HTMLButtonElement> = useRef(null);
 
-  const commonProps = {
-    id,
-    className,
-    'aria-label': accessibilityLabel,
-  };
-  const interactiveProps = {
-    ...commonProps,
-    role,
-    onClick,
-    onFocus,
-    onBlur,
-    onMouseUp: handleMouseUpByBlurring,
-    onMouseEnter,
-    onTouchStart,
-  };
+    useImperativeHandle(ref, () => {
+      return {
+        focus: () => localRef?.current?.focus(),
+      };
+    });
 
-  const handleClick = useDisableClick(disabled, onClick);
+    const commonProps = {
+      id,
+      className,
+      'aria-label': accessibilityLabel,
+    };
+    const interactiveProps = {
+      ...commonProps,
+      role,
+      onClick,
+      onFocus,
+      onBlur,
+      onMouseUp: handleMouseUpByBlurring,
+      onMouseEnter,
+      onTouchStart,
+    };
 
-  if (url) {
-    buttonMarkup = disabled ? (
-      // Render an `<a>` so toggling disabled/enabled state changes only the
-      // `href` attribute instead of replacing the whole element.
-      <a {...commonProps}>{children}</a>
-    ) : (
-      <UnstyledLink
-        {...interactiveProps}
-        url={url}
-        external={external}
-        download={download}
-        {...rest}
-      >
-        {children}
-      </UnstyledLink>
-    );
-  } else {
-    buttonMarkup = (
-      <button
-        {...interactiveProps}
-        aria-disabled={disabled}
-        type={submit ? 'submit' : 'button'}
-        aria-busy={loading ? true : undefined}
-        aria-controls={ariaControls}
-        aria-expanded={ariaExpanded}
-        aria-describedby={ariaDescribedBy}
-        aria-checked={ariaChecked}
-        aria-pressed={pressed}
-        onKeyDown={onKeyDown}
-        onKeyUp={onKeyUp}
-        onKeyPress={onKeyPress}
-        onClick={handleClick}
-        tabIndex={disabled ? -1 : undefined}
-        {...rest}
-      >
-        {children}
-      </button>
-    );
-  }
+    const handleClick = useDisableClick(disabled, onClick);
 
-  return buttonMarkup;
-}
+    if (url) {
+      buttonMarkup = disabled ? (
+        // Render an `<a>` so toggling disabled/enabled state changes only the
+        // `href` attribute instead of replacing the whole element.
+        <a {...commonProps}>{children}</a>
+      ) : (
+        <UnstyledLink
+          {...interactiveProps}
+          url={url}
+          external={external}
+          download={download}
+          {...rest}
+          ref={localRef}
```

Comment:
Do we need to have a separate ref for links since the localRef is currently only typed for HTMLButtonElement?

```suggestion
          ref={linkRef}
```